weblate/checks/source.py

```diff
@@ -112,20 +138,36 @@ def get_description(self, check_obj):
         )
 
 
-class LongUntranslatedCheck(SourceCheck):
+class LongUntranslatedCheck(SourceCheck, BatchCheckMixin):
     check_id = "long_untranslated"
     name = gettext_lazy("Long untranslated")
     description = gettext_lazy("The string has not been translated for a long time.")
 
-    def check_source_unit(self, sources: list[str], unit: Unit):
-        if unit.timestamp > timezone.now() - timedelta(days=90):
-            return False
-        states = list(unit.unit_set.values_list("state", flat=True))
-        total = len(states)
-        not_translated = states.count(STATE_EMPTY) + states.count(STATE_FUZZY)
-        translated_percent = 100 * (total - not_translated) / total
-        return (
-            total
-            and 2 * translated_percent
-            < unit.translation.component.stats.translated_percent
-        )
+    def check_source_unit(self, sources: list[str], unit: Unit) -> bool:
+        return self.handle_batch(unit, unit.translation.component)
+
+    def check_component(self, component: Component) -> Iterable[Unit]:
+        from weblate.trans.models import Unit
+
+        units = Unit.objects.filter(
+            translation__component=component,
```

Comment:
The source check should operate on source strings only:

```suggestion
            translation__component=component,
            translation_language=component.source_language,
```